File: exercises/01-sample-tracker/sample-tracker.ts
#!/usr/bin/env node

/**
 * Sample Tracker CLI
 *
 * A command-line tool for tracking laboratory samples and their pipeline
 * runs. Supports listing, searching, summarizing by status, generating
 * a subject status report, submitting new samples, and fetching sample
 * details.
 */

import * as fs from "fs";
import * as path from "path";

// Configuration
const DATA_FILE = "samples.json";

interface PipelineRun {
  id: string;
  status: string;
  pipeline_name?: string;
}

interface Sample {
  id: string;
  subject_id: string;
  barcode: string;
  sample_type: string;
  kit: string;
  status: string;
  submitted_at: string;
  completed_at?: string;
  total_reads?: number;
  pipeline_runs?: PipelineRun[];
}

function loadSamples(): any[] {
  try {
    const data = fs.readFileSync(DATA_FILE, "utf-8");
    const parsed = JSON.parse(data);
    const samples = parsed["samples"];
    console.log("Successfully loaded " + samples.length + " samples");
    return samples;
  } catch (e) {
    console.log("Error: Could not load samples file");
    console.log(e);
    return [];
  }
}

function saveSamples(samples: any[]) {
  const payload = { samples: samples };
  fs.writeFileSync(DATA_FILE, JSON.stringify(payload, null, 2));
}

function listSamples() {
  const samples = loadSamples();
  console.log("");
  console.log("=".repeat(70));
  console.log("SAMPLES");
  console.log("=".repeat(70));
  for (let i = 0; i < samples.length; i++) {
    const s = samples[i];
    let output = "";
    output = output + s["id"];
    output = output + "  " + s["subject_id"];
    output = output + "  [" + s["status"] + "]";
    output = output + "  submitted " + s["submitted_at"];
    console.log(output);
  }
  console.log("=".repeat(70));
  console.log("Total: " + samples.length + " samples");
}

function searchSamples(query: string) {
  const samples = loadSamples();
  const results: any[] = [];

  for (const s of samples) {
    if (s["subject_id"].toLowerCase().includes(query.toLowerCase())) {
      results.push(s);
    } else if (s["barcode"].toLowerCase().includes(query.toLowerCase())) {
      if (results.indexOf(s) === -1) {
        results.push(s);
      }
    }
  }

  if (results.length === 0) {
    console.log("No samples found matching '" + query + "'");
    return;
  }

  console.log("Found " + results.length + " sample(s):");
  for (const r of results) {
    console.log("  - " + r["id"] + "  " + r["subject_id"] + "  [" + r["status"] + "]");
  }
}

function summaryByStatus() {
  const samples = loadSamples();

  const statuses: string[] = [];
  for (const s of samples) {
    if (statuses.indexOf(s["status"]) === -1) {
      statuses.push(s["status"]);
    }
  }

  console.log("");
  console.log("SAMPLES BY STATUS");
  console.log("-".repeat(40));
  for (const st of statuses) {
    let count = 0;
    let totalReads = 0;
    for (const s of samples) {
      if (s["status"] === st) {
        count = count + 1;
        if (s["total_reads"]) {
          totalReads = totalReads + parseInt(s["total_reads"]);
        }
      }
    }
    console.log(st + ": " + count + " samples, " + totalReads + " total reads");
  }
}

function subjectReport(subjectId: string, results: any[] = []) {
  const samples = loadSamples();

  console.log("");
  console.log("STATUS REPORT FOR " + subjectId);
  console.log("-".repeat(60));
  for (const s of samples) {
    if (s["subject_id"] === subjectId) {
      const runs = s["pipeline_runs"] || [];

      const line = s["id"] + "  [" + s["status"] + "]  " + runs.length + " pipeline run(s)";
      console.log(line);
      results.push(s);

      for (const run of runs) {
        console.log("    - run " + run["id"] + " (" + run["status"] + ") " + (run["pipeline_name"] || ""));
      }
    }
  }

  if (results.length === 0) {
    console.log("No samples found for subject " + subjectId);
  }
}

function submitSample(subjectId: string, barcode: string, sampleType: string, kit: string) {
  const samples = loadSamples();

  const newSample: any = {};
  newSample["id"] = "smpl_" + Math.random().toString(36).substring(2, 10);
  newSample["subject_id"] = subjectId;
  newSample["barcode"] = barcode;
  newSample["sample_type"] = sampleType;
  newSample["kit"] = kit;
  newSample["status"] = "received";
  newSample["submitted_at"] = new Date().toISOString();
  newSample["pipeline_runs"] = [];

  samples.push(newSample);
  saveSamples(samples);

  console.log("Sample submitted! ID: " + newSample["id"]);
}

function getSampleDetails(sampleId: string) {
  const samples = loadSamples();

  let s: any = null;
  for (const sample of samples) {
    if (sample["id"] === sampleId) {
      s = sample;
      break;
    }
  }

  if (s === null) {
    console.log("Sample not found: " + sampleId);
    return;
  }

  console.log("ID: " + s["id"]);
  console.log("Subject: " + s["subject_id"]);
  console.log("Barcode: " + s["barcode"]);
  console.log("Type: " + s["sample_type"]);
  console.log("Kit: " + s["kit"]);
  console.log("Status: " + s["status"]);
  console.log("Submitted: " + s["submitted_at"]);
  if (s["completed_at"]) {
    console.log("Completed: " + s["completed_at"]);
  }
  if (s["total_reads"]) {
    console.log("Total Reads: " + s["total_reads"]);
  }
}

function main() {
  const args = process.argv.slice(2);

  if (args.length < 1) {
    console.log("Usage: tsx sample-tracker.ts <command> [args]");
    console.log("Commands: list, search, summary, report, submit, get");
    process.exit(1);
  }

  const command = args[0];

  if (command === "list") {
    listSamples();
  } else if (command === "search") {
    if (args.length < 2) {
      console.log("Usage: tsx sample-tracker.ts search <query>");
      process.exit(1);
    }
    searchSamples(args[1]);
  } else if (command === "summary") {
    summaryByStatus();
  } else if (command === "report") {
    if (args.length < 2) {
      console.log("Usage: tsx sample-tracker.ts report <subject_id>");
      process.exit(1);
    }
    subjectReport(args[1]);
  } else if (command === "submit") {
    if (args.length < 5) {
      console.log("Usage: tsx sample-tracker.ts submit <subject_id> <barcode> <sample_type> <kit>");
      process.exit(1);
    }
    submitSample(args[1], args[2], args[3], args[4]);
  } else if (command === "get") {
    if (args.length < 2) {
      console.log("Usage: tsx sample-tracker.ts get <sample_id>");
      process.exit(1);
    }
    getSampleDetails(args[1]);
  } else {
    console.log("Unknown command: " + command);
    process.exit(1);
  }
}

main();

File: exercises/01-sample-tracker/sample_tracker.py
#!/usr/bin/env python3
"""
Sample Tracker CLI

A command-line tool for tracking laboratory samples and their pipeline
runs. Talks to the Bioscope Samples API. Supports listing, searching,
summarizing by status, generating a subject status report, submitting
new samples, and fetching sample details.
"""

import sys
import os
import json
import requests
from datetime import datetime


# Configuration
BASE_URL = "http://localhost:8000/api/v1"
API_TOKEN = os.getenv("BIOSCOPE_API_TOKEN", "")
TIMEOUT = 30


def get_all_samples():
    """Fetch all samples from the API."""
    try:
        url = BASE_URL + "/samples"
        headers = {"Authorization": "Bearer " + API_TOKEN}
        r = requests.get(url, headers=headers, timeout=TIMEOUT)
        data = r.json()
        samples = data["samples"]
        print("Successfully fetched " + str(len(samples)) + " samples")
        return samples
    except Exception as e:
        print("Error: Could not fetch samples")
        print(e)
        return []


def list_samples():
    """List all samples in the system."""
    samples = get_all_samples()
    print("")
    print("=" * 70)
    print("SAMPLES")
    print("=" * 70)
    for i in range(0, len(samples)):
        s = samples[i]
        output = ""
        output = output + s["id"]
        output = output + "  " + s["subject_id"]
        output = output + "  [" + s["status"] + "]"
        output = output + "  submitted " + s["submitted_at"]
        print(output)
    print("=" * 70)
    print("Total: " + str(len(samples)) + " samples")


def search_samples(query):
    """Search for samples matching the query."""
    # Build the query string by hand
    url = BASE_URL + "/samples/search?q=" + query
    headers = {"Authorization": "Bearer " + API_TOKEN}
    r = requests.get(url, headers=headers)
    data = r.json()
    results = data["results"]

    if len(results) == 0:
        print("No samples found matching '" + query + "'")
        return

    print("Found " + str(len(results)) + " sample(s):")
    for r in results:
        print("  - " + r["id"] + "  " + r["subject_id"] + "  [" + r["status"] + "]")


def summary_by_status():
    """Show a summary of samples grouped by status."""
    samples = get_all_samples()

    statuses = []
    for s in samples:
        if s["status"] not in statuses:
            statuses.append(s["status"])

    print("")
    print("SAMPLES BY STATUS")
    print("-" * 40)
    for st in statuses:
        count = 0
        total_reads = 0
        for s in samples:
            if s["status"] == st:
                count = count + 1
                if s.get("total_reads"):
                    total_reads = total_reads + int(s["total_reads"])
        print(st + ": " + str(count) + " samples, " + str(total_reads) + " total reads")


def subject_report(subject_id, results=[]):
    """Generate a status report for all samples belonging to a subject."""
    samples = get_all_samples()

    print("")
    print("STATUS REPORT FOR " + subject_id)
    print("-" * 60)
    for s in samples:
        if s["subject_id"] == subject_id:
            # Fetch pipeline runs for this sample
            url = BASE_URL + "/samples/" + s["id"] + "/pipeline-runs"
            headers = {"Authorization": "Bearer " + API_TOKEN}
            r = requests.get(url, headers=headers)
            data = r.json()
            runs = data["runs"]

            line = s["id"] + "  [" + s["status"] + "]  " + str(len(runs)) + " pipeline run(s)"
            print(line)
            results.append(s)

            for run in runs:
                print("    - run " + run["id"] + " (" + run["status"] + ") " + run.get("pipeline_name", ""))

    if len(results) == 0:
        print("No samples found for subject " + subject_id)


def submit_sample(subject_id, barcode, sample_type, kit):
    """Submit a new sample to the system."""
    payload = {}
    payload["subject_id"] = subject_id
    payload["barcode"] = barcode
    payload["sample_type"] = sample_type
    payload["kit"] = kit

    url = BASE_URL + "/samples"
    headers = {"Authorization": "Bearer " + API_TOKEN, "Content-Type": "application/json"}
    r = requests.post(url, json=payload, headers=headers)

    if r.status_code == 200 or r.status_code == 201:
        sample = r.json()
        print("Sample submitted! ID: " + sample["id"])
    else:
        print("Error submitting sample: " + str(r.status_code))


def get_sample_details(sample_id):
    """Get full details for a specific sample."""
    url = BASE_URL + "/samples/" + sample_id
    headers = {"Authorization": "Bearer " + API_TOKEN}
    r = requests.get(url, headers=headers)
    s = r.json()

    print("ID: " + s["id"])
    print("Subject: " + s["subject_id"])
    print("Barcode: " + s["barcode"])
    print("Type: " + s["sample_type"])
    print("Kit: " + s["kit"])
    print("Status: " + s["status"])
    print("Submitted: " + s["submitted_at"])
    if s.get("completed_at"):
        print("Completed: " + s["completed_at"])
    if s.get("total_reads"):
        print("Total Reads: " + str(s["total_reads"]))


def main():
    """Main entry point for the application."""
    if len(sys.argv) < 2:
        print("Usage: python sample_tracker.py <command> [args]")
        print("Commands: list, search, summary, report, submit, get")
        sys.exit(1)

    command = sys.argv[1]

    if command == "list":
        list_samples()
    elif command == "search":
        if len(sys.argv) < 3:
            print("Usage: python sample_tracker.py search <query>")
            sys.exit(1)
        search_samples(sys.argv[2])
    elif command == "summary":
        summary_by_status()
    elif command == "report":
        if len(sys.argv) < 3:
            print("Usage: python sample_tracker.py report <subject_id>")
            sys.exit(1)
        subject_report(sys.argv[2])
    elif command == "submit":
        if len(sys.argv) < 6:
            print("Usage: python sample_tracker.py submit <subject_id> <barcode> <sample_type> <kit>")
            sys.exit(1)
        submit_sample(sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5])
    elif command == "get":
        if len(sys.argv) < 3:
            print("Usage: python sample_tracker.py get <sample_id>")
            sys.exit(1)
        get_sample_details(sys.argv[2])
    else:
        print("Unknown command: " + command)
        sys.exit(1)


if __name__ == "__main__":
    main()
